pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java

```diff
@@ -73,6 +74,13 @@ protected BaseCombineOperator(List<Operator> operators, QueryContext queryContex
     //       The parallelism is bounded by the task count.
     _numTasks = CombineOperatorUtils.getNumTasksForQuery(operators.size(), queryContext.getMaxExecutionThreads());
     _futures = new Future[_numTasks];
+
+    try {
+      long requestId = Long.parseLong(queryContext.getQueryOptions().get("requestId"));
+      TraceContext.register(requestId, queryContext, _futures);
```

Comment:
Would suggest not to reuse `TraceContext` and keep it just for tracing/monitoring purpose. Perhaps create sth similar to control query at runtime, e.g. `QueryRuntimeContext` to track the futures.

But in general, I'd prefer add cancel() on Operator and Plan interface to make it explicit that query is cancellable. The reason to add cancel() on Plan interface is because `queryPlan.execute()` is the entry point of query execution, so we can have `queryPlan.cancel()` to cancel a query.